Automattic/WP-Job-Manager
# Add/notify listing change

Fixes #2441 

### Changes proposed in this Pull Request

* Adds hook for updated jobs.
* ~Hooks into `wp_trash_post`~ Now covered in `post_updated` hook.
* Sends notification to `wpjobmanager.com` when job listing is changed or updated.
* Schedules job to retry 3 times if api call fails.

### Testing instructions

This is pretty manual for now.
* Set the post meta for a job as `_promoted` and set it to 1
* Update the job listing in one of these fields: `'job_title', 'job_description', 'company_name', 'application', 'job_location'`
* The API call should fail for now [ testing with wpjobmanager.com]
* Check the new option on the database: `job_manager_should_notify_promoted_jobs` to see the values set and the error message, `last_run`, `should_notify` etc
* Use a cron job plugin to check that a new job is scheduled to run: `job_manager_promoted_jobs_notification`
* If the API call succeeds, the cron job should be removed.


### New/Updated Hooks

* ~`job_manager_edit_job_listing_before_save`: hook that fires just before a job edit is persisted to the database`.

### Some Notes

- I added a new hook `job_manager_edit_job_listing_before_save` because the hook `job_manager_user_edit_job_listing` fires **after** save and by this time we don't know the exact fields that changed.
- I have set the cron schedule to `twicedaily` just as an arbitrary value. Not sure how often we would want the job to run.
- As an alternative idea to retrying 3 times, this job is not compute intensive. Also, since a notification triggers a whole fetch of the job feeds, if one update fails and the next one succeeds, there's no need to retry the first. Therefore, this implementation determines whether to retry by setting the `should_notify_jobs` option to a boolean. If one job fails, it is set to true, if another succeeds, it is set to false so there's no need to retry the first.
- There's a `get_site_id` method that calls `get_current_blog_id`. Is this okay or there's something else that we call the site ID.
- A comment in the issue says: **The trigger should be when a promoted job is removed, deactivated or updated**, I have interpreted it to mean delete and update. Anything else? What happens when the post status is changed from say 'publish' to 'pending' or 'expired'? Some sites change a post status to pending if it was edited for moderation, this is done in the settings.
